src/resolve/resolved-dns-stub.c

```diff
@@ -949,6 +949,12 @@ static void dns_stub_process_query(Manager *m, DnsStubListenerExtra *l, DnsStrea
                 return;
         }
 
+        if (set_contains(m->refuse_record_types, INT_TO_PTR(dns_question_first_key(p->question)->type))) {
+                log_debug("Got request that is refused, refusing.");
+                dns_stub_send_failure(m, l, s, p, DNS_RCODE_REFUSED, false);
+                return;
+        }
```

Comment:
so this will only refuse these look ups when done via the DNS stubs. But we also provide lookups via a dbus ipc api and via varlink ipc api. These need to refuse the requests as well.

I think it would be best to move the check into the top of dns_query_new(), and return a recognizable error there, which is then turned into DNS_RCODE_REFUSED in the stub, and into appropriate dbus/varlink errors in those ipc interfaces.

Reply:
@poettering Most of your recommendations above is implemented but i don't understand the last part about dbus/varlink errors.